test/test-parse-statement.cpp

```diff
@@ -949,6 +949,42 @@ TEST_F(test_parse_statement, switch_case_without_expression) {
   }
 }
 
+TEST_F(test_parse_statement, switch_case_with_duplicated_cases) {
+  {
+    test_parser p(u8"switch (cond) {case x: case y: case y:}"_sv,
+                  capture_diags);
+    p.parse_and_visit_statement();
+    EXPECT_THAT(p.visits, ElementsAre("visit_variable_use",       // cond
+                                      "visit_enter_block_scope",  //
+                                      "visit_variable_use",       // x
+                                      "visit_variable_use",       // y
+                                      "visit_variable_use",       // y
+                                      "visit_exit_block_scope"));
+    EXPECT_THAT(p.errors,
+                ElementsAre(DIAG_TYPE_OFFSETS(
+                    p.code, diag_duplicated_cases_in_switch_statement,
+                    duplicated_switch_case,
+                    strlen(u8"switch (cond) {case x: case y: case "), u8"y")));
+  }
+  {
+    test_parser p(
+        u8"switch (cond) {case MyEnum.A: break; case MyEnum.A: break;}"_sv,
+        capture_diags);
+    p.parse_and_visit_statement();
+    EXPECT_THAT(p.visits, ElementsAre("visit_variable_use",       // cond
+                                      "visit_enter_block_scope",  //
+                                      "visit_variable_use",       // MyEnum.A
+                                      "visit_variable_use",       // MyEnum.A
+                                      "visit_exit_block_scope"));
```

Comment:
Nit: No need to assert visits, since we're just checking for duplicated cases.
```suggestion
    EXPECT_THAT(p.errors,
                ElementsAre(DIAG_TYPE_OFFSETS(
                    p.code, diag_duplicated_cases_in_switch_statement,
                    duplicated_switch_case,
                    strlen(u8"switch (cond) {case x: case y: case "), u8"y")));
  }
  {
    test_parser p(
        u8"switch (cond) {case MyEnum.A: break; case MyEnum.A: break;}"_sv,
        capture_diags);
    p.parse_and_visit_statement();
```